classlib/src/main/java/org/teavm/classlib/java/util/TTreeMap.java

```diff
@@ -702,47 +706,39 @@ static class EntryIterator<K, V> implements TIterator<Entry<K, V>> {
 
         @Override
         public boolean hasNext() {
-            return depth > 0;
+            return path.length > 0;
         }
 
         @Override
         public Entry<K, V> next() {
             if (modCount != owner.modCount) {
                 throw new TConcurrentModificationException();
             }
-            if (depth == 0) {
+            if (path.length == 0) {
                 throw new TNoSuchElementException();
             }
-            TreeNode<K, V> node = path[--depth];
-            last = node;
-            TreeNode<K, V> down = node.down(reverse);
-            if (down != null) {
-                node = down;
-                while (node != null) {
-                    path[depth++] = node;
-                    node = node.forward(reverse);
-                }
-            }
+            last = path[path.length - 1];
+            path = owner.pathToNext(last.getKey(), reverse);
```

Comment:
It's a performance issue! You may not do so!

Reply:
It could be, but there was failure in test_navigableKeySet with IndexOutOfBoundsException.

Reply:
So that does not mean this code is wrong. Perhaps, the code that maintains TreeNode.height is wrong

Reply:
Restored back and fixed issue with removal.